File: Trees/Populating Next Right Pointers in Each Node II.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Left *Node
 *     Right *Node
 *     Next *Node
 * }
 */
 
type NodeLevel struct {
    Node  *Node
    Level int
}

func connect(root *Node) *Node {
    if root == nil {
        return root
    }

    var rightNode *Node
    bfsQueue := list.New()
    bfsQueue.PushBack(&NodeLevel{Node: root, Level: 0})
    level := 0

    for bfsQueue.Len() > 0 {
        front := bfsQueue.Front()
        bfsQueue.Remove(front)
        curr := front.Value.(*NodeLevel)

        if curr.Level == level {
            curr.Node.Next = rightNode
        } else {
            curr.Node.Next = nil
            level = curr.Level
        }

        if curr.Node.Right != nil {
            bfsQueue.PushBack(&NodeLevel{Node: curr.Node.Right, Level: level + 1})
        }
        if curr.Node.Left != nil {
            bfsQueue.PushBack(&NodeLevel{Node: curr.Node.Left, Level: level + 1})
        }        

        rightNode = curr.Node
    }

    return root
}

File: Trees/Populating Next Right Pointers in Each Node II.rs
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/

#include<queue>

class Solution {
public:
    Node* connect(Node* root) {
        if(root==nullptr) {
            return root;
        }

        queue<pair<int, Node*>> bfs_q;
        Node* right_node = nullptr;
        int level = 0;
        bfs_q.push({0, root});

        while (!bfs_q.empty()) {
            auto curr = bfs_q.front();
            bfs_q.pop();

            if (curr.first == level) {
                curr.second->next = right_node;
            } else {
                curr.second->next = nullptr;
                level = curr.first;
            }

            if (curr.second->right) {
                bfs_q.push({level+1, curr.second->right});
            }
            if (curr.second->left) {
                bfs_q.push({level+1, curr.second->left});
            }

            right_node = curr.second;
        }
        return root;
    }
};